### PR TITLE
[llvm][rtsan] Add transform pass for sanitize_realtime_unsafe

### PR DESCRIPTION
**Pull request by davidtrevelyan**
# Pass update for sanitize_realtime_unsafe

Follows https://github.com/llvm/llvm-project/pull/106754

## Motivation

Calls to system library functions such as malloc are easy for RealtimeSanitizer to intercept. If such a call is made in a `[[clang::nonblocking]]` function (a real-time context), RealtimeSanitizer will error. Real-time programmers also write their own blocking (real-time unsafe) functions that may or may not call intercepted functions. We wish to introduce a mechanism whereby RealtimeSanitizer can error on calls to these, too, if called within a real-time context.

At the same time as introducing `[[clang::nonblocking]]`, the `[[clang::blocking]]` attribute was also introduced. With the function effects warnings (as errors) activated, blocking functions cannot be called from non-blocking functions, and this is enforced at compile time. The purpose of this series of PRs is to introduce similar functionality into RealtimeSanitizer, so that it can make the equivalent check at run time.

## Implementation

We recently merged the `sanitize_realtime_unsafe` LLVM function attribute into `main`. Our next steps are to:

1. when `sanitize_realtime_unsafe` is detected, update RealtimeSanitizer's LLVM pass to insert a call to `__rtsan_notify_blocking_call`, and
2. switch on the feature by updating Clang's CodeGen to actually add the `sanitize_realtime_unsafe` attribute to the IR for functions attributed with `[[clang::blocking]]`.

Once the feature is switched on, RealtimeSanitizer will error if any calls to functions attributed with `[[clang::blocking]]` are made from `[[clang::nonblocking]]` functions.

## Integration Roadmap

The above functionality is currently split into three patches.

- [x] Introduction of the sanitize_realtime_unsafe LLVM attribute (previous PR), and
- [ ] An update to RealtimeSanitizer's LLVM pass to use it (this PR)
- [ ] An update to Clang's CodeGen to add the `sanitize_realtime_unsafe` attribute to functions attributed with `[[clang::blocking]]` (next PR)

### CODE REVIEW

**Comment by davidtrevelyan** [comment]
@cjappl for review

File: llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp [hunk 1]
```diff
@@ -45,6 +46,26 @@ static void insertCallAtAllFunctionExitPoints(Function &Fn,
         insertCallBeforeInstruction(Fn, I, InsertFnName);
 }
 
+static PreservedAnalyses rtsanPreservedCFGAnalyses() {
+  PreservedAnalyses PA;
+  PA.preserveSet<CFGAnalyses>();
+  return PA;
+}
+
+static void insertNotifyBlockingCallAtFunctionEntryPoint(Function &F) {
```
Comment (cjappl) [inline llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp:55]: In the other function, this param is called "Fn" (for better or worse, but I think it would be good to be self consistent)

File: llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp [hunk 2]
```diff
@@ -45,6 +46,26 @@ static void insertCallAtAllFunctionExitPoints(Function &Fn,
...
+  Value *NameArg = Builder.CreateGlobalString(demangle(F.getName()));
+
+  FunctionType *FuncType =
+      FunctionType::get(Type::getVoidTy(F.getContext()),
+                        {PointerType::getUnqual(F.getContext())}, false);
+
+  FunctionCallee Func = F.getParent()->getOrInsertFunction(
+      "__rtsan_notify_blocking_call", FuncType);
+
+  Builder.CreateCall(Func, {NameArg});
```
Comment (cjappl) [inline llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp:66]: Should we just combine this with a better version of `insertCallAtFunctionEntryPoint` that takes an optional args value? What are the pros/cons of that?

I know in my loop code, I will also need to insert a call here, this time with two variables to the input function. We then have two approaches:

1. Write well named helper functions that do that action and that action only (like you have here)
2. Write a general helper that takes in any number of arguments, which you construct outside the function.

Which do you feel is more appropriate?

File: llvm/test/Instrumentation/RealtimeSanitizer/rtsan_unsafe.ll [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+define void @_Z17blocking_functionv() #0 {
+  ret void
+}
+
+define noundef i32 @main() #2 {
+  call void @_Z17blocking_functionv() #4
+  ret i32 0
+}
+
+attributes #0 = { mustprogress noinline sanitize_realtime_unsafe optnone ssp uwtable(sync) }
```
Comment (cjappl) [inline llvm/test/Instrumentation/RealtimeSanitizer/rtsan_unsafe.ll:12]: And just to clarify, there is no case where someone could mark something as realtime unsafe, and realtime sanitize? (this would correspond to both [[clang::nonblocking]] and [[clang::blocking]]).

This is explicitly disallowed by the attributes right? I'm basically wondering if we need a test case here to have both attributes and ensure they are inserted in the order:

realtime_enter()
notify_blocking()
realtime_exit()

not: 
notify_blocking()
realtime_enter()
realtime_exit()

Just making sure these attributes are mutually exclusive -- if they are ignore this comment completely :)

File: llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp (hunk 2 above)
Comment (cjappl) [reply llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp:66]: In my prototype branch, you can see what the second one would look like here:

https://github.com/cjappl/llvm-project/pull/6/files#diff-a0897f69407ad91c84a9618026d8264f776688677e6a49e53164b9bf469f89afR28

I'm down for either approach, but I want to hear your thoughts on one over the other

File: llvm/test/Instrumentation/RealtimeSanitizer/rtsan_unsafe.ll [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+; RealtimeSanitizer pass should create the demangled function name as a global string, and
+; pass it as input to an inserted __rtsan_expect_not_realtime call at the function entrypoint
```
Comment (cjappl) [inline llvm/test/Instrumentation/RealtimeSanitizer/rtsan_unsafe.ll:15]: old name is in this comment.

You may consider making this comment less specific so it doesn't fall out of style as quickly "RTSan pass should insert a function call to notify the runtime of an illegal call - it passes in the demangled name of the function it is called from"

File: llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp [hunk 5]
```diff
@@ -45,6 +46,26 @@ static void insertCallAtAllFunctionExitPoints(Function &Fn,
...
 
+static PreservedAnalyses rtsanPreservedCFGAnalyses() {
+  PreservedAnalyses PA;
+  PA.preserveSet<CFGAnalyses>();
+  return PA;
+}
+
+static void insertNotifyBlockingCallAtFunctionEntryPoint(Function &F) {
+  IRBuilder<> Builder(&F.front().front());
+  Value *NameArg = Builder.CreateGlobalString(demangle(F.getName()));
```
Comment (cjappl) [inline llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp:57]: For you (or other reviewers) what is the difference between CreateGlobalString and CreateGlobalStringPtr, and which is appropriate here?

Comment (nikic) [reply llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp:57]: You should use CreateGlobalString. Deprecated in https://github.com/llvm/llvm-project/commit/04d8e364bbf758b2ecbc06b782f1b5dd91bd16e6.

File: llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp (hunk 1 above)
Comment (davidtrevelyan) [reply llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp:55]: Cool, sounds sensible - changed it to `Fn` throughout the file

File: llvm/test/Instrumentation/RealtimeSanitizer/rtsan_unsafe.ll (hunk 3 above)
Comment (davidtrevelyan) [reply llvm/test/Instrumentation/RealtimeSanitizer/rtsan_unsafe.ll:12]: Yeah good question - the attributes are indeed mutually exclusive. You can see the test in `main` here: https://github.com/llvm/llvm-project/blob/main/llvm/test/Verifier/rtsan-attrs.ll

File: llvm/test/Instrumentation/RealtimeSanitizer/rtsan_unsafe.ll (hunk 4 above)
Comment (davidtrevelyan) [reply llvm/test/Instrumentation/RealtimeSanitizer/rtsan_unsafe.ll:15]: Nice idea. Thanks

File: llvm/test/Instrumentation/RealtimeSanitizer/rtsan_unsafe.ll (hunk 3 above)
Comment (cjappl) [reply llvm/test/Instrumentation/RealtimeSanitizer/rtsan_unsafe.ll:12]: Thanks for confirming

File: llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp (hunk 2 above)
Comment (davidtrevelyan) [reply llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp:66]: Thanks for the suggestion. I was originally intending to keep the diff minimal, but I prefer refactoring for code re-use. I've pushed up a change that I think tidies it up nicely - let me know what you think!

File: llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp [hunk 6]
```diff
@@ -45,6 +46,26 @@ static void insertCallAtAllFunctionExitPoints(Function &Fn,
         insertCallBeforeInstruction(Fn, I, InsertFnName);
 }
 
+static PreservedAnalyses rtsanPreservedCFGAnalyses() {
+  PreservedAnalyses PA;
+  PA.preserveSet<CFGAnalyses>();
+  return PA;
+}
```
[the remark this answers is not in the document]
Comment (davidtrevelyan) [reply llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp:66]: I personally like the D.R.Y.ness of it, but happy to change it if you prefer!

File: llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp [hunk 7]
```diff
@@ -17,47 +17,77 @@
...
+#include "llvm/Demangle/Demangle.h"
 #include "llvm/Transforms/Instrumentation/RealtimeSanitizer.h"
 
+#include <vector>
+
 using namespace llvm;
 
+static std::vector<Type *> getArgTypes(ArrayRef<Value *> FunctionArgs) {
+  std::vector<Type *> Types;
+  for (Value *Arg : FunctionArgs)
```
Comment (cjappl) [inline llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp:29]: Could reserve here before the for loop to prevent multiple resizing

**Review by cjappl — APPROVED** [review]
Looks good to me. Let's wait to see if there is any other feedback, otherwise I will merge next week

File: llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp (hunk 7 above)
Comment (davidtrevelyan) [reply llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp:30]: Done

File: llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp [hunk 8]
```diff
@@ -17,47 +17,78 @@
...
 
+#include "llvm/Demangle/Demangle.h"
 #include "llvm/Transforms/Instrumentation/RealtimeSanitizer.h"
 
+#include <vector>
+
 using namespace llvm;
 
+static std::vector<Type *> getArgTypes(ArrayRef<Value *> FunctionArgs) {
+  std::vector<Type *> Types;
```
Comment (nikic) [inline llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp:28]: Prefer SmallVector instead of std::vector, even if you don't want to use inline storage.

File: llvm/test/Instrumentation/RealtimeSanitizer/rtsan_unsafe.ll [hunk 9]
```diff
@@ -0,0 +1,19 @@
+; RUN: opt < %s -passes=rtsan -S | FileCheck %s
```
Comment (nikic) [inline llvm/test/Instrumentation/RealtimeSanitizer/rtsan_unsafe.ll:1]: Use update_test_checks.py, unless there's some reason this is not possible here.

File: llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp (hunk 8 above)
Comment (davidtrevelyan) [reply llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp]: Thank you!

File: llvm/test/Instrumentation/RealtimeSanitizer/rtsan_unsafe.ll (hunk 9 above)
Comment (davidtrevelyan) [reply llvm/test/Instrumentation/RealtimeSanitizer/rtsan_unsafe.ll:2]: Awesome, great tool. Thanks for showing me this script - I've updated this test now

File: llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp [hunk 10]
```diff
@@ -17,47 +17,75 @@
...
+
+static PreservedAnalyses runSanitizeRealtime(Function &Fn) {
+  insertCallAtFunctionEntryPoint(Fn, "__rtsan_realtime_enter", {});
+  insertCallAtAllFunctionExitPoints(Fn, "__rtsan_realtime_exit", {});
+  return rtsanPreservedCFGAnalyses();
+}
+
+static PreservedAnalyses runSanitizeRealtimeUnsafe(Function &Fn) {
+  IRBuilder<> Builder(&Fn.front().front());
+  Value *Name = Builder.CreateGlobalString(demangle(Fn.getName()));
```
Comment (fmayer) [inline llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp:74]: Mostly out of interest, why do we de-mangle here and not in the reporting?

File: llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp [hunk 11]
```diff
@@ -17,47 +17,75 @@
...
+                                           ArrayRef<Value *> FunctionArgs) {
+  insertCallBeforeInstruction(Fn, Fn.front().front(), InsertFnName,
+                              FunctionArgs);
 }
 
 static void insertCallAtAllFunctionExitPoints(Function &Fn,
-                                              const char *InsertFnName) {
+                                              const char *InsertFnName,
+                                              ArrayRef<Value *> FunctionArgs) {
   for (auto &BB : Fn)
```
Comment (fmayer) [inline llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp:54]: driveby: you can make this nicer

```
for (auto &I : instructions(F)) {
[...]
}
```

Comment (davidtrevelyan) [reply llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp]: lovely - thanks - updated in https://github.com/llvm/llvm-project/pull/109543/commits/b4db72675b6ba8a50492c912fca8349021205784

File: llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp (hunk 10 above)
Comment (davidtrevelyan) [reply llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp:74]: That's a really great question. In all honesty I didn't originally think about the possibility of de-mangling during the report stage. After thinking about it further, I have an idea on which one will probably be less work overall, but I'd need advice on whether the distribution of work is appropriate.

Here's how I see the work involved:

- for demangling in the pass:
   - no run-time work: demangles once at compile time, but
   - could increase compile-time work (marginally?)
 - for demangling in the reporting:
   - adds to run-time work: demangles every time an error is registered (which could be many times if `halt_on_error` is off)
   - won't add to compile-time work 
   
 My very naive intuition is that demangling in the pass is preferable, but very happy to move it to the reporting if I have my priorities a bit upside down. Thanks in advance for any suggestions here.

Comment (fmayer) [reply llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp:74]: I think you are missing the downside of binary size, because demangled name takes up more space, for demangling in the pass.

I am not sure if demangling will cause any noticable overhead over unwinding _and_ symbolizing the stack frame.

Taking a step back, why do we pass this name and not use the PC and symbolize that?

Comment (davidtrevelyan) [reply llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp:74]: Thanks for this question and prompting us to look into this in more detail. I had indeed not considered the issue of binary size, so went away and thought about it a bit harder with @cjappl. There's lots to think about here, especially in terms of how we expect the sanitizer to be used.

We anticipate that the sanitizer users will _only be marking a handful of functions they are worried about_ with the `[[clang::blocking]]` attribute, which is what ultimately results in `sanitize_realtime_unsafe` being added in IR. We think this because any functions that do the normal bad real-time stuff like allocation will be picked up by the libc interceptors anyway. Adding the `[[clang::blocking]]` attribute is more for user functions that could block _entirely within user space_, like a spin mutex lock or similar busy-wait scenario. Conservatively estimating that a user has 10 concerning `[[blocking]]` functions, each with demangled symbol length 25 characters - I think that would be (at least ideally) of the order 250 bytes of extra binary size to store. The rtsan runtime library is (when stripped) currently about 750 kB already, so we think this is about a 0.03% size increment - and only for sanitized builds.

We tried the PC approach in a branch and confirmed it works, but we were sad to lose the function name for printing diagnostics in release builds, and also to have to do the extra run-time work to unwind the stack. Being able to run the sanitizer with release builds is one of our design goals - we'd like to achieve as-near-normal performance on the real-time thread as possible, so that users can meaningfully QA their real-time systems with RealtimeSanitizer enabled.

Given:

- the extra work at run time necessary to unwind the stack,
- the simpler run-time implementation when accepting a function name string,
- the loss of being able to print the blocking function name in release builds, and
- the almost-negligible expected binary increment,

... our moderate preference is currently to keep the approach how we have it outlined in this PR. However, this is of course caveated with the appreciation that we're very new here and inexperienced with the codebase, so I'm very keen to take as much advice as possible on this. Many thanks for your help so far - keen to hear your thoughts!

**Comment by davidtrevelyan** [comment]
Polite ping @vitalybuka @MaskRay :)

File: llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp (hunk 10 above)
Comment (fmayer) [reply llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp:74]: > We tried the PC approach in a branch and confirmed it works, but we were sad to lose the function name for printing diagnostics in release builds

Confirming that this also meant that the whole stack trace symbolization was gone, but that was okay.

> do the extra run-time work to unwind the stack. 

I don't understand. To symbolize a PC, we don't need to unwind the stack. From the runtime CLs I saw go buy, it seems like you unwind the stack on error regardless though

**Review by fmayer — APPROVED** [review]

**Review by MaskRay — APPROVED** [review]

File: llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp (hunk 10 above)
Comment (davidtrevelyan) [reply llvm/lib/Transforms/Instrumentation/RealtimeSanitizer.cpp:74]: > Confirming that this also meant that the whole stack trace symbolization was gone, but that was okay.

Yes - thanks - that was indeed my intended meaning. 

> I don't understand. To symbolize a PC, we don't need to unwind the stack. From the runtime CLs I saw go buy, it seems like you unwind the stack on error regardless though

That's correct, my mistake. We can discount the runtime cost of unwinding the stack and just concentrate on the functionality/binary size trade-off. After some more thought, we also realised that the symbolised PC approach would make suppression of these errors difficult to implement, because we would no longer have a function name to match against a suppression list. For this reason, we're leaning more strongly towards the existing implmentation and taking the small hit on binary size. I noticed that you approved this PR last night, so I guess you're happy enough with it as it is - please let me know if you have any further concerns. Thanks again for the review 👍
